### PR TITLE
Track type information for IR expressions

### PR DESCRIPTION
**Pull request by gsomix**
For now pointer member access operator (#129) works only for lvalues, but not rvalues. E.g. this sample doesn't compile:
```c
foo x;
(&x)->x = 42; // error
```
In #129 to find members by identifier we relied on `GetValueType: unit -> TypeReference` method implemented for all lvalues. I propose to extend this approach and track type information for all IR expressions:
```c
foo x;
// `x` is lvalue expression of type `foo`
// `(&x)` is rvalue expression of type `foo*`
// `(&x)->x` is lvalue expression of type `int`
(&x)->x = 42; // works!
(42)->x = 42; // error: "Int32" has no member named "x"
```

- [x] Fix parser for expressions in parentheses
- [x] Implement `GetConstantType` for constants
- [x] Implement `GetExpessionType` for address-of operator (`&`) expression
- [x] Implement `GetExpressionType` for other unary expressions
- ~Implement `GetExpessionType` for binary operator expressions~ See #136 
- [x] Implement `GetExpressionType` for other expressions
- [x] Fix pointer member access operator
- [x] Tests, tests, tests!
- [x] Add sample

### CODE REVIEW

File: Cesium.CodeGen/Ir/Expressions/BinaryOperatorExpression.cs [hunk 1]
```diff
@@ -65,6 +66,9 @@ public virtual void EmitTo(IDeclarationScope scope)
         };
     }
 
+    // TODO[139]: Implement conversions and types tracking for arithmetic operations
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Ir/Expressions/BinaryOperatorExpression.cs:69]: Is it correct that it talks about #139? Probably not? Did you wanted to reference #136 here?

File: Cesium.CodeGen/Ir/Expressions/Constants/StringConstant.cs [hunk 2]
```diff
@@ -21,4 +22,7 @@ public void EmitTo(IDeclarationScope scope)
         var fieldReference = scope.AssemblyContext.GetConstantPoolReference(_value);
         scope.Method.Body.Instructions.Add(Instruction.Create(OpCodes.Ldsflda, fieldReference));
     }
+
+    // TODO: String? Not a char* or something?
+    public TypeReference GetConstantType(IDeclarationScope scope) => scope.TypeSystem.String;
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Ir/Expressions/Constants/StringConstant.cs:27]: Should be `char*`, yeah.

File: Cesium.Parser/CParser.cs [hunk 3]
```diff
@@ -68,6 +68,9 @@ public partial class CParser
     private static Expression MakeStringLiteralExpression(ICToken stringLiteral) =>
         new ConstantExpression(stringLiteral);
 
+    [Rule("primary_expression: '(' expression ')'")]
+    private static Expression MakeParens(IToken _, Expression expression, IToken __) => expression;
+
     // TODO:
     // primary-expression:
     //     ( expression )
```
Comment (ForNeVeR) [inline Cesium.Parser/CParser.cs:76]: Please delete this TODO line, since you've implemented this expression already.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]
In general, I approve this plan. There are some small issues, though.

**Review by ForNeVeR — APPROVED** [review]
Thanks!

**Comment by gsomix** [comment]
Thanks for the review!

**Pull request merged**